src/workers/parser/getOutOfScopeLocations.js

```diff
@@ -38,26 +38,26 @@ function getLocation(func) {
 }
 
 /**
- * Reduces an array of locations to remove items that are completely enclosed
- * by another location in the array.
+ * Return an new locations array which excludes
+ * items that are completely enclosed by another location in the input locations
+ *
+ * @param locations Notice! The locations MUST be sorted by `sortByStart`
+ *                  so that we can do linear time complexity operation.
  */
-function removeOverlaps(
-  locations: AstLocation | AstLocation[],
-  location: AstLocation
-) {
-  // support reducing without an initializing array
-  if (!Array.isArray(locations)) {
-    locations = [locations];
+function removeOverlaps(locations: AstLocation[]) {
+  const newLocs = [];
+  const length = locations.length;
+  if (length) {
+    newLocs.push(locations[0]);
+    for (let i = 1, parentIdx = 0; i < length; i++) {
+      const loc = locations[i];
+      if (containsLocation(locations[parentIdx], loc) == false) {
+        newLocs.push(loc);
+        parentIdx = i;
+      }
+    }
   }
-
-  const contains =
-    locations.filter(a => containsLocation(a, location)).length > 0;
```

Comment:
This is O(n) complexity.
Consider there are 10 locations to test: [ L1, L2, ..., L10 ] and only L7 contains L8 ~ L10.
Loop 1: we get [ L1 ]
Loop 2: we test L2 against [ L1 ]
Loop 3: we test L3 against [ L1, L2 ]
... ...
Loop 8: we test L8 against [ L1, .., L7 ] 
Loop 9: we test L9 against [ L1, .., L7 ] 
Loop 10: we test L10 against [ L1, .., L7 ]